### ISSUE TITLE
conflict hex value display between disassembly-flavor and disasm result in Arch Linux

### ISSUE THREAD

**Issue by RocketMaDev** [pwndbg/pwndbg#2324]
### Description

I know the default disassembly-flavor for pwndbg is intel, and code should be displayed like `sub rsp, 0x10`. But it's so weird on my system that these hex values are displayed like `10h`. An example screenshot is attached below.

![屏幕截图_20240727_220718](https://github.com/user-attachments/assets/0dfb78e9-0ddb-4ebd-bd35-b48487080552)

I tried to `set disassembly-flavor intel`, nothing happened. But when I `set disassembly-flavor att`, the hex value displayed become normal like:

![屏幕截图_20240727_220901](https://github.com/user-attachments/assets/c8a9098c-2deb-44a2-9951-010e24cf049c)

Even more weirdly, I then `set disassembly-flavor intel` back, and the code was displayed just like the screenshot above, it didn't go back to `sub rsp, 10h`.

### Steps to reproduce

My `.gdbinit`:

```gdbinit
set debuginfod enabled on
source /usr/share/pwndbg/gdbinit.py
add-auto-load-safe-path /usr/lib/go/src/runtime/runtime-gdb.py
set context-sections regs disasm code backtrace threads expressions
```

Setting disassembly-flavor to intel and att some times when debugging or write it into gdbinit

### My setup

System: Arch Linux
Gdb:      15.1
Python:   3.12.4 (main, Jun  7 2024, 06:33:07) [GCC 14.1.1 20240522]
Pwndbg:   2024.02.14
Capstone: 5.0.1280
Unicorn:  2.0.1

**Comment by disconnect3d** [comment]
Hi, this is ... interesting. We use capstone for disassembling code. I really wonder why this behaves as you described.

Do you maybe want to try to debug this? Otherwise I can look at it in upcomming days.

**Comment by RocketMaDev** [comment]
Actually I haven't dug into capstone. I'm just a pwndbg user. I can take a look at it if you could tell me where to check the code. Or else I can wait for your update.

**Comment by OBarronCS** [comment]
This is a very strange bug. Capstone is providing the hex literal, and these lines initialize the disassembler:

https://github.com/pwndbg/pwndbg/blob/0f61bff5bb290279eefb61a5f8c616c7da51b82d/pwndbg/gdblib/disasm/__init__.py#L138-L151

**Comment by disconnect3d** [comment]
Hmm I can't reproduce this on my VPS though it also doesn't have emulation due to too low RAM memory...

**Comment by disconnect3d** [comment]
@RocketMaDev We still don't know why this occurs on your side. Any chance you can experiment a bit with this and see if u find the cause?

Some ideas:
- run pwndbg with .gdbinit that only instantiates pwndbg
- run gdb+pwndbg without any environmennt variables
- dig into our code where we disassembly code and see where does this come from

**Comment by RocketMaDev** [comment]
Ok, I will try to explore.

**Comment by RocketMaDev** [comment]
That is interesting. I found a weird behavior: Capstone don't accept CS_OPT_SYNTAX_ATT
![屏幕截图_20240817_112058](https://github.com/user-attachments/assets/496d2149-63a2-4968-b6f9-61ea11349948)

**Comment by RocketMaDev** [comment]
Maybe it is not caused by cs.syntax? I did some logging, and found that `cs.syntax` remains unchanged, but the output of DISASM section DO changes.
![屏幕截图_20240817_113230](https://github.com/user-attachments/assets/799e51d5-2dfe-43ab-8102-aacf2ae82911)
![屏幕截图_20240817_112901](https://github.com/user-attachments/assets/8a7b1fed-89e9-40ae-be9d-d51920d2cd44)

**Comment by OBarronCS** [comment]
I tested this same thing on my machine - for whatever reason, the `except` path is not being taken for me - ie CS_OPT_SYNTAX_ATT is accepted on my local machine. Very odd...

I dug a bit into the Capstone code for this, and found the underlying handler for setting `md.syntax`:

https://github.com/capstone-engine/capstone/blob/5e6807bab9ba6a8112b8ba1f92285924e699ee84/arch/X86/X86Module.c#L55

Maybe somehow the the number constants do not line up, or could it possibly somehow be compiled differently on your machine?

**Comment by RocketMaDev** [comment]
Guys, what's your Capstone version?

**Comment by OBarronCS** [comment]
I have the same version as your output above - 5.0.1280

**Comment by RocketMaDev** [comment]
According Capstone doc, `csh` in Cs is the handle to native Capstone object. Looking through the log, when set flavor to att, the handle keeps freshing while setting flavor to intel keeps handle unchanged.

![屏幕截图_20240817_164142](https://github.com/user-attachments/assets/278354c3-b5e1-471c-a97d-a2aecca0e78b)

```log
2024-08-17 16:37:25,297 - DEBUG - intel
2024-08-17 16:37:25,297 - DEBUG - 4
2024-08-17 16:37:25,297 - DEBUG - 4
2024-08-17 16:37:25,298 - DEBUG - <CsInsn 0x555555555149 [55]: push rbp>
2024-08-17 16:37:25,298 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,299 - DEBUG - <CsInsn 0x5555555550f0 [f30f1efa]: endbr64 >
2024-08-17 16:37:25,299 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,300 - DEBUG - <CsInsn 0x7ffff7fcb200 [f30f1efa]: endbr64 >
2024-08-17 16:37:25,300 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,301 - DEBUG - <CsInsn 0x55555555514d [4883ec10]: sub rsp, 10h>
2024-08-17 16:37:25,301 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,305 - DEBUG - <CsInsn 0x555555555151 [bf00000000]: mov edi, 0>
2024-08-17 16:37:25,305 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,305 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 555555555040h>
2024-08-17 16:37:25,305 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,306 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:37:25,306 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,306 - DEBUG - <CsInsn 0x55555555515f [488d059e0e0000]: lea rax, [rip + 0e9eh]>
2024-08-17 16:37:25,306 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,306 - DEBUG - <CsInsn 0x555555555166 [4889c7]: mov rdi, rax>
2024-08-17 16:37:25,306 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,307 - DEBUG - <CsInsn 0x555555555169 [e8c2feffff]: call 555555555030h>
2024-08-17 16:37:25,307 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,307 - DEBUG - <CsInsn 0x55555555516e [b800000000]: mov eax, 0>
2024-08-17 16:37:25,307 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,307 - DEBUG - <CsInsn 0x555555555173 [c9]: leave >
2024-08-17 16:37:25,307 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,307 - DEBUG - <CsInsn 0x555555555174 [c3]: ret >
2024-08-17 16:37:25,307 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:25,308 - DEBUG - <CsInsn 0x555555555175 [0000]: add byte ptr [rax], al>
2024-08-17 16:37:25,308 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,803 - DEBUG - <CsInsn 0x555555555149 [55]: push rbp>
2024-08-17 16:37:38,803 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,804 - DEBUG - <CsInsn 0x5555555550f0 [f30f1efa]: endbr64 >
2024-08-17 16:37:38,804 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,804 - DEBUG - <CsInsn 0x7ffff7fcb200 [f30f1efa]: endbr64 >
2024-08-17 16:37:38,805 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,806 - DEBUG - <CsInsn 0x555555555151 [bf00000000]: mov edi, 0>
2024-08-17 16:37:38,806 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,806 - DEBUG - <CsInsn 0x55555555514d [4883ec10]: sub rsp, 10h>
2024-08-17 16:37:38,806 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,808 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 555555555040h>
2024-08-17 16:37:38,808 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,808 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:37:38,808 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,808 - DEBUG - <CsInsn 0x55555555515f [488d059e0e0000]: lea rax, [rip + 0e9eh]>
2024-08-17 16:37:38,808 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x555555555166 [4889c7]: mov rdi, rax>
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x555555555169 [e8c2feffff]: call 555555555030h>
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x55555555516e [b800000000]: mov eax, 0>
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x555555555173 [c9]: leave >
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x555555555174 [c3]: ret >
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:38,809 - DEBUG - <CsInsn 0x555555555175 [0000]: add byte ptr [rax], al>
2024-08-17 16:37:38,809 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:53,398 - DEBUG - <CsInsn 0x555555555149 [55]: push rbp>
2024-08-17 16:37:53,398 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:53,398 - DEBUG - <CsInsn 0x5555555550f0 [f30f1efa]: endbr64 >
2024-08-17 16:37:53,399 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:53,399 - DEBUG - <CsInsn 0x7ffff7fcb200 [f30f1efa]: endbr64 >
2024-08-17 16:37:53,399 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:53,400 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 555555555040h>
2024-08-17 16:37:53,400 - DEBUG - c_ulong(106501415483584)
2024-08-17 16:37:53,401 - DEBUG - att
2024-08-17 16:37:53,401 - INFO - Invalid option (CS_ERR_OPTION)
2024-08-17 16:37:53,401 - DEBUG - 4
2024-08-17 16:37:53,401 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 0x555555555040>
2024-08-17 16:37:53,401 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,402 - DEBUG - <CsInsn 0x555555555151 [bf00000000]: mov edi, 0>
2024-08-17 16:37:53,402 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,402 - DEBUG - <CsInsn 0x55555555514d [4883ec10]: sub rsp, 0x10>
2024-08-17 16:37:53,402 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,405 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:37:53,405 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,405 - DEBUG - <CsInsn 0x55555555515f [488d059e0e0000]: lea rax, [rip + 0xe9e]>
2024-08-17 16:37:53,405 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,405 - DEBUG - <CsInsn 0x555555555166 [4889c7]: mov rdi, rax>
2024-08-17 16:37:53,405 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,405 - DEBUG - <CsInsn 0x555555555169 [e8c2feffff]: call 0x555555555030>
2024-08-17 16:37:53,405 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,406 - DEBUG - <CsInsn 0x55555555516e [b800000000]: mov eax, 0>
2024-08-17 16:37:53,406 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,406 - DEBUG - <CsInsn 0x555555555173 [c9]: leave >
2024-08-17 16:37:53,406 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,406 - DEBUG - <CsInsn 0x555555555174 [c3]: ret >
2024-08-17 16:37:53,406 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:53,406 - DEBUG - <CsInsn 0x555555555175 [0000]: add byte ptr [rax], al>
2024-08-17 16:37:53,406 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:58,780 - DEBUG - <CsInsn 0x5555555550f0 [f30f1efa]: endbr64 >
2024-08-17 16:37:58,780 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:58,781 - DEBUG - <CsInsn 0x7ffff7fcb200 [f30f1efa]: endbr64 >
2024-08-17 16:37:58,781 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:58,781 - DEBUG - <CsInsn 0x7ffff7fc5ac0 [f30f1efa]: endbr64 >
2024-08-17 16:37:58,781 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:58,782 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:37:58,782 - DEBUG - c_ulong(106501404133872)
2024-08-17 16:37:58,783 - DEBUG - att
2024-08-17 16:37:58,783 - INFO - Invalid option (CS_ERR_OPTION)
2024-08-17 16:37:58,783 - DEBUG - 4
2024-08-17 16:37:58,783 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:37:58,783 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,784 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 0x555555555040>
2024-08-17 16:37:58,784 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,784 - DEBUG - <CsInsn 0x555555555151 [bf00000000]: mov edi, 0>
2024-08-17 16:37:58,784 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,785 - DEBUG - <CsInsn 0x55555555514d [4883ec10]: sub rsp, 0x10>
2024-08-17 16:37:58,785 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,787 - DEBUG - <CsInsn 0x55555555515f [488d059e0e0000]: lea rax, [rip + 0xe9e]>
2024-08-17 16:37:58,787 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,787 - DEBUG - <CsInsn 0x555555555166 [4889c7]: mov rdi, rax>
2024-08-17 16:37:58,787 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,787 - DEBUG - <CsInsn 0x555555555169 [e8c2feffff]: call 0x555555555030>
2024-08-17 16:37:58,787 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,788 - DEBUG - <CsInsn 0x55555555516e [b800000000]: mov eax, 0>
2024-08-17 16:37:58,788 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,788 - DEBUG - <CsInsn 0x555555555173 [c9]: leave >
2024-08-17 16:37:58,788 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,788 - DEBUG - <CsInsn 0x555555555174 [c3]: ret >
2024-08-17 16:37:58,788 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:37:58,788 - DEBUG - <CsInsn 0x555555555175 [0000]: add byte ptr [rax], al>
2024-08-17 16:37:58,788 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,878 - DEBUG - <CsInsn 0x5555555550f0 [f30f1efa]: endbr64 >
2024-08-17 16:38:07,878 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,879 - DEBUG - <CsInsn 0x7ffff7fcb200 [f30f1efa]: endbr64 >
2024-08-17 16:38:07,879 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,880 - DEBUG - <CsInsn 0x7ffff7fc5ac0 [f30f1efa]: endbr64 >
2024-08-17 16:38:07,880 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,881 - DEBUG - <CsInsn 0x55555555515f [488d059e0e0000]: lea rax, [rip + 0xe9e]>
2024-08-17 16:38:07,881 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,882 - DEBUG - <CsInsn 0x55555555515b [488945f8]: mov qword ptr [rbp - 8], rax>
2024-08-17 16:38:07,882 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,882 - DEBUG - <CsInsn 0x555555555156 [e8e5feffff]: call 0x555555555040>
2024-08-17 16:38:07,882 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,882 - DEBUG - <CsInsn 0x555555555151 [bf00000000]: mov edi, 0>
2024-08-17 16:38:07,882 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,882 - DEBUG - <CsInsn 0x55555555514d [4883ec10]: sub rsp, 0x10>
2024-08-17 16:38:07,882 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,884 - DEBUG - <CsInsn 0x555555555166 [4889c7]: mov rdi, rax>
2024-08-17 16:38:07,884 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,885 - DEBUG - <CsInsn 0x555555555169 [e8c2feffff]: call 0x555555555030>
2024-08-17 16:38:07,885 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,885 - DEBUG - <CsInsn 0x55555555516e [b800000000]: mov eax, 0>
2024-08-17 16:38:07,885 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,885 - DEBUG - <CsInsn 0x555555555173 [c9]: leave >
2024-08-17 16:38:07,885 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,885 - DEBUG - <CsInsn 0x555555555174 [c3]: ret >
2024-08-17 16:38:07,885 - DEBUG - c_ulong(106501416468320)
2024-08-17 16:38:07,886 - DEBUG - <CsInsn 0x555555555175 [0000]: add byte ptr [rax], al>
2024-08-17 16:38:07,886 - DEBUG - c_ulong(106501416468320)
```

**Comment by RocketMaDev** [comment]
I comment out these lines:

```python
    try:
        flavor = gdb.execute("show disassembly-flavor", to_string=True).lower().split('"')[1]
        logging.debug(flavor) # remove me
    except gdb.error as e:
        logging.info(e) # remove me
        if str(e).find("disassembly-flavor") > -1:
            flavor = "intel"
        else:
            raise

    cs = Cs(arch, mode)
    # try:
    #     cs.syntax = CapstoneSyntax[flavor]
    #     logging.debug(cs.syntax) # remove me
    # except CsError as ex:
    #     logging.info(ex) # remove me
    #     pass
    cs.detail = True
    logging.debug(cs.syntax) # remove me
    return cs
```

now it works. 
![屏幕截图_20240817_173439](https://github.com/user-attachments/assets/65e43e24-dd7e-4d63-b152-3e4669bb9555)

**Comment by disconnect3d** [comment]
Well I don't think we should comment this code because then we won't be able to switch the capstone syntax if I understand correctly? :)

From the code that @OBarronCS linked, it seems that the ATT syntax may be disabled/removed from capstone when you build it:
```c
				case CS_OPT_SYNTAX_ATT:
#if !defined(CAPSTONE_DIET) && !defined(CAPSTONE_X86_ATT_DISABLE)
					handle->printer = X86_ATT_printInst;
					handle->syntax = CS_OPT_SYNTAX_ATT;
					break;
```

Could it be that this is the case on your installation? How did you install Capstone?

I wonder if we can reliably detect it and inform the user somehow etc.

**Comment by RocketMaDev** [comment]
@OBarronCS Yes, you are right. I dug into capstone library itself, and found that there is a confusion in `CS_OPT_SYNTAX_*` macros. Their values are not synchronized between Python binding and C include files.

In detail, this bug was introduced in v5.0.1 and is fixed in v5.0.2. In `next` branch, the macro value updated, but they weren't introduced into stable branch. However, Python binding updated through cherry-pick, which caused this bug. More details here: capstone-engine/capstone#2240

@disconnect3d I'm sorry that this is a bug in Capstone and is not a bug of pwndbg. Either fallback to 4.x or upgrade to 5.0.2 fix the bug.

For users using Arch Linux like me, the fix has been available via pacman. Just run `sudo pacman -Syu`.
